06x-new-kafka-roller.md

```diff
@@ -0,0 +1,326 @@
+
+#### Quorum health check
+
+The quorum health logic is similar to the current KafkaRoller except for a couple of differences. The current KafkaRoller uses the `controller.quorum.fetch.timeout.ms` config value from the desired configurations passed from the reconciler or uses the hard-coded default value if the reconciler pass null for desired configurations. The new roller will use the configuration value of the active controller. This will mean that the quorum health check is done from the active controller's point of view.
+
+Also the current KafkaRoller does not connect to the controller via Admin API to get the quorum health information. By the time, we implement this proposal, Strimzi should support Kafka 3.7 which includes [KIP 919](https://cwiki.apache.org/confluence/display/KAFKA/KIP-919%3A+Allow+AdminClient+to+Talk+Directly+with+the+KRaft+Controller+Quorum+and+add+Controller+Registration). Therefore new KafkaRoller will be able to connect to the controller directly for quorum information and active controller's configuration.
+
+#### Availability check
+
+The availibility check logic similar to the current KafkaRoller. The ISRs that the broker is part of is checked against the configured under minimum ISR size. If `size(ISR containing the broker) - minISR > 0`, the broker can be considered safe to restart. If it equals to 0, restarting the broker could cause under minimum ISR partition. If it's less than 0, it means the partition is already under minimum ISR and restarting it would either not make a difference or make things worse. In both cases, the broker should not be restarted. 
+
+However, if `size(Replicas containing the broker) - minISR <= 0` but the topic partition is configured with replication size less than minISR, the check will pass to proceed with the broker restart.
+
+#### An example of rolling update
+
+Here is an example of the new roller performing rolling restarts on a cluster with 12 nodes: 3 controllers, 3 mixed nodes and 6 brokers. The nodes are:
+- controller-0
+- controller-1
+- controller-2
+- mixed-3
+- mixed-4
+- mixed-5
+- broker-6
+- broker-7
+- broker-8
+- broker-9
+- broker-10
+- broker-11
+
+1. The roller observes nodes and update their contexts based on the observation outcome:
+
+All the nodes except `mixed-3` have the following Context with `nodeRef` being their `podname/node-id`, and `nodeRoles` having either `controller`, `broker` or both.
+```
+         nodeRef: controller-0/0
+         nodeRoles: controller
+         state: READY
+         lastTransition: 0123456
+         restartReason: MANUAL_ROLLING_UPDATE
+         numRestartAttempts: 0
+         numReconfigAttempts: 0
+         numRetries: 0
+```
+The `mixed-3` node has the following context because the operator could not establish an admin connection to it even though it's ready from Kubernetes and KafkaAgent perspective:
+```
+         nodeRef: mixed-3/3
+         nodeRoles: controller,broker
+         state: NOT_RUNNING
```

Comment:
Shouldn't this be UNKNOWN from the point of view of the operator? I mean, unless it was not running before and then became unresponsive. Maybe adding the initial state in the node list at the start of the example would help. Something like:

```sh
- controller-0 (ready)
...
- mixed-3 (not running, unresponsive)
...
```

Reply:
you are right, I wouldn't expect to be checked for responsiveness if it was not running. I think it should be NOT_READY or READY. UNKNOWN is if we didn't observe its state yet. We would check the state and then connection.